### ISSUE TITLE
Question about parse-server-mailgun...

### ISSUE THREAD

**Issue by LilMoke** [sebsylvester/parse-server-mailgun#41]
Hello,

I have set up a parse server and need to send mail.  I installed https://github.com/parse-community/parse-server-simple-mailgun-adapter and https://github.com/mailgun/mailgun-js

BUT, I was wondering if your parse-server-mailgun can be used in place of those?

I am pretty new to parse but have a fairly good understanding.  I just wanted to know if you could answer a few questions:

1) If I use your code do I need mailgun-js and/or the parse adaptor?
2) Will this work with the Parse password reset, or do I need to provide my own call to send a password reset email
3) Do I just need to install your parse-server-mailgun and mustache?  What mailgun items need to be installed?

Thank you os much for any help you can give me!!!

Best,
Tony

**Comment by LilMoke** [comment]
Ok, I got your adapter set up and it appears to sort of be working but I get this error:

{"code":1,"message":"Internal server error."}

Any ideas?  Could it me on the mailgun side?  

I was not sure what to put here:

const MailgunAdapter = AppCache.get('yourAppId').userController.adapter;

Is this the same as in my index.js file?

**Comment by sebsylvester** [comment]
Replace the placeholder "yourAppId" with same value you used when configuring the Parse Server, like in this [example](https://github.com/parse-community/parse-server#parse-server--express).

**Comment by LilMoke** [comment]
Here is my cloud function, is this correct?

```
Parse.Cloud.define("sendEmail", function(request, response) {
    // Get access to Parse Server's cache
    const {
        AppCache
    } = require('parse-server/lib/cache');
    // Get a reference to the MailgunAdapter
    // NOTE: It's best to do this inside the Parse.Cloud.define(...) method body and not at the top of your file with your other imports. This gives Parse Server time to boot, setup cloud code and the email adapter.
    const MailgunAdapter = AppCache.get('KEY').userController.adapter;

    // Invoke the send method with an options object
    MailgunAdapter.send({
        templateName: 'custom_email',
        // Optional override of your configuration's subject
        subject: 'Important: action required',
        // Optional override of the adapter's fromAddress
		fromAddress: 'Tony <xxxx@xxx.com>',
        recipient: 'xxx@xxx.com',
        variables: {
            alert: 'New posts'
        }, // {{alert}} will be compiled to 'New posts'
        // Additional message fields can be included with the "extra" option
        // See https://nodemailer.com/extras/mailcomposer/#e-mail-message-fields for an overview of what can be included
        extra: {
            attachments: [ /* include attachment objects */ ],
            replyTo: 'reply-to-address'
        }
	}, {
			success: function() {
				response.success("Mail sent ok");
			},
			error: function(error) {
				response.error("Mail send error: " + error);
		}
	});
});

```

**Comment by LilMoke** [comment]
Here is the message I receive:

```
verbose: REQUEST for [POST] /parse/functions/sendEmail: {
  "message": "Hello there!"
} method=POST, url=/parse/functions/sendEmail, host=xxxx, x-real-ip=184.209.82.7, x-forwarded-for=xxx, content-length=26, content-type=application/json; charset=utf-8, accept=*/*, x-parse-session-token=r:e6721d34d06fa56a6e0ee085756aa2bf, x-parse-application-id=xxxx, x-parse-client-key=xxx, x-parse-installation-id=fc8224d2-c6fe-4be6-85b0-3dee862a19c3, x-parse-os-version=11.3 (15E216), accept-language=en-us, accept-encoding=gzip, deflate, x-parse-client-version=i1.17.1, user-agent=Streamlist-iOS/40 CFNetwork/897.15 Darwin/17.5.0, x-parse-app-build-version=40, x-parse-app-display-version=1.1, message=Hello there!
error: Error generating response. TypeError: MailgunAdapter.send is not a function
    at /var/app/current/cloud/main.js:50:20
    at /var/app/current/node_modules/parse-server/lib/Routers/FunctionsRouter.js:176:9
    at new Promise (<anonymous>)
    at handleCloudFunction (/var/app/current/node_modules/parse-server/lib/Routers/FunctionsRouter.js:144:14)
    at /var/app/current/node_modules/parse-server/lib/PromiseRouter.js:169:7
    at Layer.handle [as handle_request] (/var/app/current/node_modules/express/lib/router/layer.js:95:5)
    at next (/var/app/current/node_modules/express/lib/router/route.js:137:13)
    at Route.dispatch (/var/app/current/node_modules/express/lib/router/route.js:112:3)
    at Layer.handle [as handle_request] (/var/app/current/node_modules/express/lib/router/layer.js:95:5)
    at /var/app/current/node_modules/express/lib/router/index.js:281:22 

```

I removed the success and error functions but still get this error.  How do I get the response from Mailgun and return it from my cloud func?

Thank you for the help!!!

**Comment by sebsylvester** [comment]
The ``send`` method returns a Promise object, so you could attach ``then`` and ``catch`` handlers to it in order to evaluate the response from Mailgun. Use the ``response`` argument from the cloud function to send stuff back to the caller. This should be well documented in the Parse Server docs

**Comment by LilMoke** [comment]
Ok, thanks!  I will look at the Parse Server Doc, but I think I know how to do that.

SO, does my cloud function look ok other than the response stuff?  

I really do not know why I am getting:

`MailgunAdapter.send is not a function
`

I am VERY new to .js so any help would be greatly appreciated!!!

Best,
T

**Comment by LilMoke** [comment]
So I have put then and catch handlers but I still get:

`MailgunAdapter.send is not a function
`
I am really not sure why I get this error?

**Comment by LilMoke** [comment]
Hmmm, i think my package may not be right.  

this is the package.json i use to reply this to my AWS:

```
{
  "name": "parse-server",
  "version": "2.7.4",
  "description": "Parse API server using the parse-server module",
  "main": "index.js",
	"repository": {
		"type": "git",
		"url": "https://github.com/parse-community/parse-server"
	},
  "license": "BSD-3-Clause",
  "dependencies": {
    "express": "4.16.2",
    "kerberos": "0.0.x",
	"parse": "~1.11.1",
	"parse-server": "~2.7.4",
	"parse-dashboard": "~1.2.0",
	"parse-server-mailgun": "2.4.8"
},
	"scripts": {
		"start": "node index.js"
	},
	"engines": {
		"node": ">=6.11.4"
	}
}

```

I also run the this from your GitHub:

`npm install --save parse-server-mailgun
`

Is that correct?

**Comment by LilMoke** [comment]
So, where is the email template directory supposed to be?  It never finds the template... also, should the template have .txt or .html at the end?  I don't think so , but it cannot find it.

**Comment by sebsylvester** [comment]
The[ docs demonstrate](https://github.com/sebsylvester/parse-server-mailgun#adapter-configuration) how to set up your template directory. Take a closer look at the ``pathPlainText`` and ``pathHtml`` properties.

**Comment by LilMoke** [comment]
Yes, got that. Thank you for the help.  I found the problem.  It was my misunderstanding of exactly how it worked.  When I looked at your MailgunAdapter.js source I realized my foolish mistake.

I did not realize that the templates in the index.js file were the templates it was searching for in the cloud function so I had the names mismatched.  I thought the index.js was only for Parse to find emails it needed, such as Password Reset.

Now I understand and everything is working perfectly.

Thanks again for all your help!!
